File: app/src/main/java/org/thoughtcrime/securesms/components/AudioView.java
package org.thoughtcrime.securesms.components;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.Rect;
import android.net.Uri;
import android.util.AttributeSet;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.graphics.drawable.DrawableCompat;
import androidx.lifecycle.Observer;

import com.airbnb.lottie.LottieAnimationView;
import com.airbnb.lottie.LottieProperty;
import com.airbnb.lottie.SimpleColorFilter;
import com.airbnb.lottie.model.KeyPath;
import com.airbnb.lottie.value.LottieValueCallback;
import com.pnikosis.materialishprogress.ProgressWheel;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.signal.core.util.logging.Log;
import org.thoughtcrime.securesms.R;
import org.thoughtcrime.securesms.audio.AudioWaveForm;
import org.thoughtcrime.securesms.components.voice.VoiceNotePlaybackState;
import org.thoughtcrime.securesms.database.AttachmentTable;
import org.thoughtcrime.securesms.events.PartProgressEvent;
import org.thoughtcrime.securesms.mms.AudioSlide;
import org.thoughtcrime.securesms.mms.SlideClickListener;

import java.util.Objects;
import java.util.concurrent.TimeUnit;

public final class AudioView extends FrameLayout {

  private static final String TAG = Log.tag(AudioView.class);

  private static final int MODE_NORMAL = 0;
  private static final int MODE_SMALL  = 1;
  private static final int MODE_DRAFT  = 2;

  private static final int FORWARDS =  1;
  private static final int REVERSE  = -1;

  @NonNull  private final AnimatingToggle     controlToggle;
  @NonNull  private final View                progressAndPlay;
  @NonNull  private final LottieAnimationView playPauseButton;
  @NonNull  private final ImageView           downloadButton;
  @Nullable private final ProgressWheel       circleProgress;
  @NonNull  private final SeekBar             seekBar;
            private final boolean             smallView;
            private final boolean             autoRewind;

  @Nullable private final TextView duration;

  @ColorInt private final int waveFormPlayedBarsColor;
  @ColorInt private final int waveFormUnplayedBarsColor;
  @ColorInt private final int waveFormThumbTint;

  @Nullable private SlideClickListener downloadListener;
            private int                backwardsCounter;
            private int                lottieDirection;
            private boolean            isPlaying;
            private long               durationMillis;
            private AudioSlide         audioSlide;
            private Callbacks          callbacks;

  private final Observer<VoiceNotePlaybackState> playbackStateObserver = this::onPlaybackState;

  public AudioView(Context context) {
    this(context, null);
  }

  public AudioView(Context context, AttributeSet attrs) {
    this(context, attrs, 0);
  }

  public AudioView(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
    setLayoutDirection(LAYOUT_DIRECTION_LTR);

    TypedArray typedArray = null;
    try {
      typedArray = context.getTheme().obtainStyledAttributes(attrs, R.styleable.AudioView, 0, 0);

      int mode   = typedArray.getInteger(R.styleable.AudioView_audioView_mode, MODE_NORMAL);
      smallView  = mode == MODE_SMALL;
      autoRewind = typedArray.getBoolean(R.styleable.AudioView_autoRewind, false);

      switch (mode) {
        case MODE_NORMAL:
          inflate(context, R.layout.audio_view, this);
          break;
        case MODE_SMALL:
          inflate(context, R.layout.audio_view_small, this);
          break;
        case MODE_DRAFT:
          inflate(context, R.layout.audio_view_draft, this);
          break;
        default:
          throw new IllegalStateException("Unsupported mode: " + mode);
      }

      this.controlToggle   = findViewById(R.id.control_toggle);
      this.playPauseButton = findViewById(R.id.play);
      this.progressAndPlay = findViewById(R.id.progress_and_play);
      this.downloadButton  = findViewById(R.id.download);
      this.circleProgress  = findViewById(R.id.circle_progress);
      this.seekBar         = findViewById(R.id.seek);
      this.duration        = findViewById(R.id.duration);

      lottieDirection = REVERSE;
      this.playPauseButton.setOnClickListener(new PlayPauseClickedListener());
      this.playPauseButton.setOnLongClickListener(v -> performLongClick());
      this.seekBar.setOnSeekBarChangeListener(new SeekBarModifiedListener());

      setTint(typedArray.getColor(R.styleable.AudioView_foregroundTintColor, Color.WHITE));

      int backgroundTintColor = typedArray.getColor(R.styleable.AudioView_backgroundTintColor, Color.TRANSPARENT);
      if (getBackground() != null && backgroundTintColor != Color.TRANSPARENT) {
        DrawableCompat.setTint(getBackground(), backgroundTintColor);
      }

      this.waveFormPlayedBarsColor   = typedArray.getColor(R.styleable.AudioView_waveformPlayedBarsColor, Color.WHITE);
      this.waveFormUnplayedBarsColor = typedArray.getColor(R.styleable.AudioView_waveformUnplayedBarsColor, Color.WHITE);
      this.waveFormThumbTint         = typedArray.getColor(R.styleable.AudioView_waveformThumbTint, Color.WHITE);

      setProgressAndPlayBackgroundTint(typedArray.getColor(R.styleable.AudioView_progressAndPlayTint, Color.BLACK));
    } finally {
      if (typedArray != null) {
        typedArray.recycle();
      }
    }
  }

  @Override
  protected void onAttachedToWindow() {
    super.onAttachedToWindow();
    if (!EventBus.getDefault().isRegistered(this)) EventBus.getDefault().register(this);
  }

  @Override
  protected void onDetachedFromWindow() {
    super.onDetachedFromWindow();
    EventBus.getDefault().unregister(this);
  }

  public void setProgressAndPlayBackgroundTint(@ColorInt int color) {
    progressAndPlay.getBackground().setColorFilter(color, PorterDuff.Mode.SRC_IN);
  }

  public Observer<VoiceNotePlaybackState> getPlaybackStateObserver() {
    return playbackStateObserver;
  }

  public void setAudio(final @NonNull AudioSlide audio,
                       final @Nullable Callbacks callbacks,
                       final boolean showControls,
                       final boolean forceHideDuration)
  {
    this.callbacks = callbacks;

    if (duration != null) {
      duration.setVisibility(View.VISIBLE);
    }

    if (seekBar instanceof WaveFormSeekBarView) {
      if (audioSlide != null && !Objects.equals(audioSlide.getUri(), audio.getUri())) {
       WaveFormSeekBarView waveFormView = (WaveFormSeekBarView) seekBar;
       waveFormView.setWaveMode(false);
       seekBar.setProgress(0);
       durationMillis = 0;
      }
    }

    if (showControls && audio.isPendingDownload()) {
      controlToggle.displayQuick(downloadButton);
      seekBar.setEnabled(false);
      downloadButton.setOnClickListener(new DownloadClickedListener(audio));
      if (circleProgress != null) {
        if (circleProgress.isSpinning()) circleProgress.stopSpinning();
        circleProgress.setVisibility(View.GONE);
      }
    } else if (showControls && audio.getTransferState() == AttachmentTable.TRANSFER_PROGRESS_STARTED) {
      controlToggle.displayQuick(progressAndPlay);
      seekBar.setEnabled(false);
      if (circleProgress != null) {
        circleProgress.setVisibility(View.VISIBLE);
        circleProgress.spin();
      }
    } else {
      seekBar.setEnabled(true);
      if (circleProgress != null && circleProgress.isSpinning()) circleProgress.stopSpinning();
      showPlayButton();
    }

    this.audioSlide = audio;

    if (seekBar instanceof WaveFormSeekBarView) {
      WaveFormSeekBarView waveFormView = (WaveFormSeekBarView) seekBar;
      waveFormView.setColors(waveFormPlayedBarsColor, waveFormUnplayedBarsColor, waveFormThumbTint);
      if (android.os.Build.VERSION.SDK_INT >= 23) {
        new AudioWaveForm(getContext(), audio).getWaveForm(
          data -> {
            durationMillis = data.getDuration(TimeUnit.MILLISECONDS);
            updateProgress(0, 0);
            if (!forceHideDuration && duration != null) {
              duration.setVisibility(VISIBLE);
            }
            waveFormView.setWaveData(data.getWaveForm());
          },
          () -> waveFormView.setWaveMode(false));
      } else {
        waveFormView.setWaveMode(false);
        if (duration != null) {
          duration.setVisibility(GONE);
        }
      }
    }

    if (forceHideDuration && duration != null) {
      duration.setVisibility(View.GONE);
    }
  }

  public void setDownloadClickListener(@Nullable SlideClickListener listener) {
    this.downloadListener = listener;
  }

  public @Nullable Uri getAudioSlideUri() {
    if (audioSlide != null) return audioSlide.getUri();
    else                    return null;
  }

  private void onPlaybackState(@NonNull VoiceNotePlaybackState voiceNotePlaybackState) {
    onDuration(voiceNotePlaybackState.getUri(), voiceNotePlaybackState.getTrackDuration());
    onProgress(voiceNotePlaybackState.getUri(),
               (double) voiceNotePlaybackState.getPlayheadPositionMillis() / voiceNotePlaybackState.getTrackDuration(),
               voiceNotePlaybackState.getPlayheadPositionMillis());
    onSpeedChanged(voiceNotePlaybackState.getUri(), voiceNotePlaybackState.getSpeed());
    onStart(voiceNotePlaybackState.getUri(), voiceNotePlaybackState.isPlaying(), voiceNotePlaybackState.isAutoReset());
  }

  private void onDuration(@NonNull Uri uri, long durationMillis) {
    if (isTarget(uri)) {
      this.durationMillis = durationMillis;
    }
  }

  private void onStart(@NonNull Uri uri, boolean statePlaying, boolean autoReset) {
    if (!isTarget(uri) || !statePlaying) {
      if (hasAudioUri()) {
        onStop(audioSlide.getUri(), autoReset);
      }

      return;
    }

    if (isPlaying) {
      return;
    }

    isPlaying = true;
    togglePlayToPause();
  }

  private void onStop(@NonNull Uri uri, boolean autoReset) {
    if (!isTarget(uri)) {
      return;
    }

    if (!isPlaying) {
      return;
    }

    isPlaying = false;
    togglePauseToPlay();

    if (autoReset || autoRewind || seekBar.getProgress() + 5 >= seekBar.getMax()) {
      backwardsCounter = 4;
      rewind();
    }
  }

  private void onProgress(@NonNull Uri uri, double progress, long millis) {
    if (!isTarget(uri)) {
      return;
    }

    int seekProgress = (int) Math.floor(progress * seekBar.getMax());

    if (seekProgress > seekBar.getProgress() || backwardsCounter > 3) {
      backwardsCounter = 0;
      seekBar.setProgress(seekProgress);
      updateProgress((float) progress, millis);
    } else {
      backwardsCounter++;
    }
  }

  private void onSpeedChanged(@NonNull Uri uri, float speed) {
    if (callbacks != null) {
      callbacks.onSpeedChanged(speed, isTarget(uri));
    }
  }

  private boolean isTarget(@NonNull Uri uri) {
    return hasAudioUri() && Objects.equals(uri, audioSlide.getUri());
  }

  private boolean hasAudioUri() {
    return audioSlide != null && audioSlide.getUri() != null;
  }

  @Override
  public void setFocusable(boolean focusable) {
    super.setFocusable(focusable);
    this.playPauseButton.setFocusable(focusable);
    this.seekBar.setFocusable(focusable);
    this.seekBar.setFocusableInTouchMode(focusable);
    this.downloadButton.setFocusable(focusable);
  }

  @Override
  public void setClickable(boolean clickable) {
    super.setClickable(clickable);
    this.playPauseButton.setClickable(clickable);
    this.seekBar.setClickable(clickable);
    this.seekBar.setOnTouchListener(clickable ? new LongTapAwareTouchListener() : new TouchIgnoringListener());
    this.downloadButton.setClickable(clickable);
  }

  @Override
  public void setEnabled(boolean enabled) {
    super.setEnabled(enabled);
    this.playPauseButton.setEnabled(enabled);
    this.seekBar.setEnabled(enabled);
    this.downloadButton.setEnabled(enabled);
  }

  private void updateProgress(float progress, long millis) {
    if (callbacks != null) {
      callbacks.onProgressUpdated(durationMillis, millis);
    }

    if (duration != null && durationMillis > 0) {
      long remainingSecs = Math.max(0, TimeUnit.MILLISECONDS.toSeconds(durationMillis - millis));
      duration.setText(getResources().getString(R.string.AudioView_duration, remainingSecs / 60, remainingSecs % 60));
    }

    if (smallView && circleProgress != null) {
      circleProgress.setInstantProgress(seekBar.getProgress() == 0 ? 1 : progress);
    }
  }

  public void setTint(int foregroundTint) {
    post(()-> this.playPauseButton.addValueCallback(new KeyPath("**"),
                                                    LottieProperty.COLOR_FILTER,
                                                    new LottieValueCallback<>(new SimpleColorFilter(foregroundTint))));

    this.downloadButton.setColorFilter(foregroundTint, PorterDuff.Mode.SRC_IN);

    if (circleProgress != null) {
      this.circleProgress.setBarColor(foregroundTint);
    }

    if (this.duration != null) {
      this.duration.setTextColor(foregroundTint);
    }
    this.seekBar.getProgressDrawable().setColorFilter(foregroundTint, PorterDuff.Mode.SRC_IN);
    this.seekBar.getThumb().setColorFilter(foregroundTint, PorterDuff.Mode.SRC_IN);
  }

  public void getSeekBarGlobalVisibleRect(@NonNull Rect rect) {
    seekBar.getGlobalVisibleRect(rect);
  }

  private double getProgress() {
    if (this.seekBar.getProgress() <= 0 || this.seekBar.getMax() <= 0) {
      return 0;
    } else {
      return (double)this.seekBar.getProgress() / (double)this.seekBar.getMax();
    }
  }

  private void togglePlayToPause() {
    startLottieAnimation(FORWARDS);
  }

  private void togglePauseToPlay() {
    startLottieAnimation(REVERSE);
  }

  private void startLottieAnimation(int direction) {
    showPlayButton();

    if (lottieDirection == direction) {
      return;
    }
    lottieDirection = direction;

    playPauseButton.pauseAnimation();
    playPauseButton.setSpeed(direction * 2);
    playPauseButton.resumeAnimation();
  }

  private void showPlayButton() {
    if (circleProgress != null) {
      if (!smallView) {
        circleProgress.setVisibility(GONE);
      } else if (seekBar.getProgress() == 0) {
        circleProgress.setInstantProgress(1);
      }
    }

    playPauseButton.setVisibility(VISIBLE);
    controlToggle.displayQuick(progressAndPlay);
  }

  public void stopPlaybackAndReset() {
    if (audioSlide == null || audioSlide.getUri() == null) return;

    if (callbacks != null) {
      callbacks.onStopAndReset(audioSlide.getUri());
      rewind();
    }
  }

  private class PlayPauseClickedListener implements View.OnClickListener {

    @Override
    public void onClick(View v) {
      if (audioSlide == null || audioSlide.getUri() == null) return;

      if (callbacks != null) {
        if (lottieDirection == REVERSE) {
          callbacks.onPlay(audioSlide.getUri(), getProgress());
        } else {
          callbacks.onPause(audioSlide.getUri());
        }
      }
    }
  }

  private void rewind() {
    seekBar.setProgress(0);
    updateProgress(0, 0);
  }

  private class DownloadClickedListener implements View.OnClickListener {
    private final @NonNull AudioSlide slide;

    private DownloadClickedListener(@NonNull AudioSlide slide) {
      this.slide = slide;
    }

    @Override
    public void onClick(View v) {
      if (downloadListener != null) downloadListener.onClick(v, slide);
    }
  }

  private class SeekBarModifiedListener implements SeekBar.OnSeekBarChangeListener {

    private boolean wasPlaying;

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
    }

    @Override
    public synchronized void onStartTrackingTouch(SeekBar seekBar) {
      if (audioSlide == null || audioSlide.getUri() == null) return;

      wasPlaying = isPlaying;
      if (isPlaying) {
        if (callbacks != null) {
          callbacks.onPause(audioSlide.getUri());
        }
      }
    }

    @Override
    public synchronized void onStopTrackingTouch(SeekBar seekBar) {
      if (audioSlide == null || audioSlide.getUri() == null) return;

      if (callbacks != null) {
        if (wasPlaying) {
          callbacks.onSeekTo(audioSlide.getUri(), getProgress());
        } else {
          callbacks.onProgressUpdated(durationMillis, Math.round(durationMillis * getProgress()));
        }
      }
    }
  }

  private class LongTapAwareTouchListener implements OnTouchListener {
    private final GestureDetector gestureDetector = new GestureDetector(AudioView.this.getContext(), new GestureDetector.SimpleOnGestureListener() {
      @Override
      public void onLongPress(MotionEvent e) {
        performLongClick();
      }
    });

    @Override
    public boolean onTouch(View v, MotionEvent event) {
      return gestureDetector.onTouchEvent(event);
    }
  }

  private static class TouchIgnoringListener implements OnTouchListener {
    @Override
    public boolean onTouch(View v, MotionEvent event) {
      return true;
    }
  }

  @Subscribe(sticky = true, threadMode = ThreadMode.MAIN)
  public void onEventAsync(final PartProgressEvent event) {
    if (audioSlide != null && circleProgress != null && event.attachment.equals(audioSlide.asAttachment())) {
      circleProgress.setInstantProgress(((float) event.progress) / event.total);
    }
  }

  public interface Callbacks {
    void onPlay(@NonNull Uri audioUri, double progress);
    void onPause(@NonNull Uri audioUri);
    void onSeekTo(@NonNull Uri audioUri, double progress);
    void onStopAndReset(@NonNull Uri audioUri);
    void onSpeedChanged(float speed, boolean isPlaying);
    void onProgressUpdated(long durationMillis, long playheadMillis);
  }
}

File: src/org/thoughtcrime/securesms/components/AudioView.kt
package org.thoughtcrime.securesms.components

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.drawable.AnimatedVectorDrawable
import android.os.Build
import android.util.AttributeSet
import android.view.View
import android.view.View.OnTouchListener
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.graphics.BlendModeColorFilterCompat.createBlendModeColorFilterCompat
import androidx.core.graphics.BlendModeCompat
import com.pnikosis.materialishprogress.ProgressWheel
import network.loki.messenger.R
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import org.thoughtcrime.securesms.audio.AudioSlidePlayer
import org.thoughtcrime.securesms.database.AttachmentDatabase
import org.thoughtcrime.securesms.events.PartProgressEvent
import org.thoughtcrime.securesms.logging.Log
import org.thoughtcrime.securesms.mms.AudioSlide
import org.thoughtcrime.securesms.mms.SlideClickListener
import java.io.IOException
import java.util.*
import java.util.concurrent.TimeUnit
import kotlin.math.floor

class AudioView: FrameLayout, AudioSlidePlayer.Listener {

    companion object {
        private const val TAG = "AudioViewKt"
    }

    private val controlToggle: AnimatingToggle
    private val container: ViewGroup
    private val playButton: ImageView
    private val pauseButton: ImageView
    private val downloadButton: ImageView
    private val downloadProgress: ProgressWheel
    private val seekBar: WaveformSeekBar
    private val timestamp: TextView

    private var downloadListener: SlideClickListener? = null
    private var audioSlidePlayer: AudioSlidePlayer? = null
    private var backwardsCounter = 0

    constructor(context: Context): this(context, null)

    constructor(context: Context, attrs: AttributeSet?): this(context, attrs, 0)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int): super(context, attrs, defStyleAttr) {
        View.inflate(context, R.layout.audio_view, this)
        container = findViewById(R.id.audio_widget_container)
        controlToggle = findViewById(R.id.control_toggle)
        playButton = findViewById(R.id.play)
        pauseButton = findViewById(R.id.pause)
        downloadButton = findViewById(R.id.download)
        downloadProgress = findViewById(R.id.download_progress)
        seekBar = findViewById(R.id.seek)
        timestamp = findViewById(R.id.timestamp)

        playButton.setOnClickListener {
            try {
                Log.d(TAG, "playbutton onClick")
                if (audioSlidePlayer != null) {
                    togglePlayToPause()

                    // Restart the playback if progress bar is near at the end.
                    val progress = if (seekBar.progress < 0.99f) seekBar.progress.toDouble() else 0.0

                    audioSlidePlayer!!.play(progress)
                }
            } catch (e: IOException) {
                Log.w(TAG, e)
            }
        }
        pauseButton.setOnClickListener {
            Log.d(TAG, "pausebutton onClick")
            if (audioSlidePlayer != null) {
                togglePauseToPlay()
                audioSlidePlayer!!.stop()
            }
        }
        seekBar.progressChangeListener = object : WaveformSeekBar.ProgressChangeListener {
            override fun onProgressChanged(waveformSeekBar: WaveformSeekBar, progress: Float, fromUser: Boolean) {
                if (fromUser && audioSlidePlayer != null) {
                    synchronized(audioSlidePlayer!!) {
                        audioSlidePlayer!!.seekTo(progress.toDouble())
                    }
                }
            }
        }
        //TODO Remove this.
        seekBar.sample = Random().let { (0 until 64).map { i -> it.nextFloat() }.toFloatArray() }

        playButton.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.play_icon))
        pauseButton.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.pause_icon))
        playButton.background = ContextCompat.getDrawable(context, R.drawable.ic_circle_fill_white_48dp)
        pauseButton.background = ContextCompat.getDrawable(context, R.drawable.ic_circle_fill_white_48dp)

        if (attrs != null) {
            val typedArray = context.theme.obtainStyledAttributes(attrs, R.styleable.AudioView, 0, 0)
            setTint(typedArray.getColor(R.styleable.AudioView_foregroundTintColor, Color.WHITE),
                    typedArray.getColor(R.styleable.AudioView_backgroundTintColor, Color.WHITE))
            container.setBackgroundColor(typedArray.getColor(R.styleable.AudioView_widgetBackground, Color.TRANSPARENT))
            typedArray.recycle()
        }
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (!EventBus.getDefault().isRegistered(this)) EventBus.getDefault().register(this)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        EventBus.getDefault().unregister(this)
    }

    fun setAudio(audio: AudioSlide, showControls: Boolean) {
        if (showControls && audio.isPendingDownload) {
            controlToggle.displayQuick(downloadButton)
            seekBar.isEnabled = false
            downloadButton.setOnClickListener { v -> downloadListener?.onClick(v, audio) }
            if (downloadProgress.isSpinning) {
                downloadProgress.stopSpinning()
            }
        } else if (showControls && audio.transferState == AttachmentDatabase.TRANSFER_PROGRESS_STARTED) {
            controlToggle.displayQuick(downloadProgress)
            seekBar.isEnabled = false
            downloadProgress.spin()
        } else {
            controlToggle.displayQuick(playButton)
            seekBar.isEnabled = true
            if (downloadProgress.isSpinning) {
                downloadProgress.stopSpinning()
            }
        }
        audioSlidePlayer = AudioSlidePlayer.createFor(context, audio, this)
    }

    fun cleanup() {
        if (audioSlidePlayer != null && pauseButton.visibility == View.VISIBLE) {
            audioSlidePlayer!!.stop()
        }
    }

    fun setDownloadClickListener(listener: SlideClickListener?) {
        downloadListener = listener
    }

    fun setTint(foregroundTint: Int, backgroundTint: Int) {
        playButton.backgroundTintList = ColorStateList.valueOf(foregroundTint)
        playButton.imageTintList = ColorStateList.valueOf(backgroundTint)
        pauseButton.backgroundTintList = ColorStateList.valueOf(foregroundTint)
        pauseButton.imageTintList = ColorStateList.valueOf(backgroundTint)

        downloadButton.setColorFilter(foregroundTint, PorterDuff.Mode.SRC_IN)
        downloadProgress.barColor = foregroundTint
        timestamp.setTextColor(foregroundTint)

//        val colorFilter = createBlendModeColorFilterCompat(foregroundTint, BlendModeCompat.SRC_IN)
//        seekBar.progressDrawable.colorFilter = colorFilter
//        seekBar.thumb.colorFilter = colorFilter
    }

    override fun onPlayerStart(player: AudioSlidePlayer) {
        if (pauseButton.visibility != View.VISIBLE) {
            togglePlayToPause()
        }
    }

    override fun onPlayerStop(player: AudioSlidePlayer) {
        if (playButton.visibility != View.VISIBLE) {
            togglePauseToPlay()
        }

//        if (seekBar.progress + 5 >= seekBar.max) {
//            backwardsCounter = 4
//            onProgress(0.0, 0)
//        }
    }

    override fun onPlayerProgress(player: AudioSlidePlayer, progress: Double, millis: Long) {
//        val seekProgress = floor(progress * seekBar.max).toInt()
        //TODO Update text.
        seekBar.progress = progress.toFloat()
//        if (/*seekProgress > 1f || */backwardsCounter > 3) {
//            backwardsCounter = 0
//            seekBar.progress = 1f
//            timestamp.text = String.format("%02d:%02d",
//                    TimeUnit.MILLISECONDS.toMinutes(millis),
//                    TimeUnit.MILLISECONDS.toSeconds(millis))
//        } else {
//            backwardsCounter++
//        }
    }

    override fun setFocusable(focusable: Boolean) {
        super.setFocusable(focusable)
        playButton.isFocusable = focusable
        pauseButton.isFocusable = focusable
        seekBar.isFocusable = focusable
        seekBar.isFocusableInTouchMode = focusable
        downloadButton.isFocusable = focusable
    }

    override fun setClickable(clickable: Boolean) {
        super.setClickable(clickable)
        playButton.isClickable = clickable
        pauseButton.isClickable = clickable
        seekBar.isClickable = clickable
        seekBar.setOnTouchListener(if (clickable) null else
            OnTouchListener { _, _ -> return@OnTouchListener true }) // Suppress touch events.
        downloadButton.isClickable = clickable
    }

    override fun setEnabled(enabled: Boolean) {
        super.setEnabled(enabled)
        playButton.isEnabled = enabled
        pauseButton.isEnabled = enabled
        seekBar.isEnabled = enabled
        downloadButton.isEnabled = enabled
    }

    private fun togglePlayToPause() {
        controlToggle.displayQuick(pauseButton)
        val playToPauseDrawable = ContextCompat.getDrawable(context, R.drawable.play_to_pause_animation) as AnimatedVectorDrawable
        pauseButton.setImageDrawable(playToPauseDrawable)
        playToPauseDrawable.start()
    }

    private fun togglePauseToPlay() {
        controlToggle.displayQuick(playButton)
        val pauseToPlayDrawable = ContextCompat.getDrawable(context, R.drawable.pause_to_play_animation) as AnimatedVectorDrawable
        playButton.setImageDrawable(pauseToPlayDrawable)
        pauseToPlayDrawable.start()
    }

//    private inner class SeekBarModifiedListener : OnSeekBarChangeListener {
//        override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {}
//
//        @Synchronized
//        override fun onStartTrackingTouch(seekBar: SeekBar) {
//            if (audioSlidePlayer != null && pauseButton.visibility == View.VISIBLE) {
//                audioSlidePlayer!!.stop()
//            }
//        }
//
//        @Synchronized
//        override fun onStopTrackingTouch(seekBar: SeekBar) {
//            try {
//                if (audioSlidePlayer != null && pauseButton.visibility == View.VISIBLE) {
//                    audioSlidePlayer!!.play(getProgress())
//                }
//            } catch (e: IOException) {
//                Log.w(TAG, e)
//            }
//        }
//    }

    @Subscribe(sticky = true, threadMode = ThreadMode.MAIN)
    fun onEventAsync(event: PartProgressEvent) {
        if (audioSlidePlayer != null && event.attachment == audioSlidePlayer!!.audioSlide.asAttachment()) {
            downloadProgress.setInstantProgress(event.progress.toFloat() / event.total)
        }
    }
}